WreckEngine/UI.cpp

```diff
@@ -0,0 +1,402 @@
+#include "UI.h"
+#include "gl_structs.h"
+#include "GLError.h"
+#include "External.h"
+#if defined(_WIN32) || defined(_WIN64)
+#define GLFW_EXPOSE_NATIVE_WIN32
+#include <GLFW/glfw3native.h>
+#endif
+
+#if defined(offsetof)
+#define OFFSETOF(TYPE, ELEMENT) offsetof(TYPE, ELEMENT)
+#else
+#define OFFSETOF(TYPE, ELEMENT) ((size_t)&(((TYPE *)0)->ELEMENT))
+#endif
+
+namespace UI
+{
+    // NOTE - Shaders and globals are copied from ImGui's GLFW example
+
+    static const char* const ImGuiVertexShader =
+        "#version 330\n"
+        "uniform mat4 ProjMtx;\n"
+        "in vec2 Position;\n"
+        "in vec2 UV;\n"
+        "in vec4 Color;\n"
+        "out vec2 Frag_UV;\n"
+        "out vec4 Frag_Color;\n"
+        "void main()\n"
+        "{\n"
+        "   Frag_UV = UV;\n"
+        "   Frag_Color = Color;\n"
+        "   gl_Position = ProjMtx * vec4(Position.xy,0,1);\n"
+        "}\n";
+
+    static const char* const ImGuiFragmentShader =
+        "#version 330\n"
+        "uniform sampler2D Texture;\n"
+        "in vec2 Frag_UV;\n"
+        "in vec4 Frag_Color;\n"
+        "out vec4 Out_Color;\n"
+        "void main()\n"
+        "{\n"
+        "   Out_Color = Frag_Color * texture( Texture, Frag_UV.st);\n"
+        "}\n";
+
+    static double   g_Time = 0.0;
```

Comment:
I always use name prefixes. And the problem with using those for input is multiple events could overwrite values in a single frame. That's why I wanted to GLFW event delegating.